components/push/src/communications.rs

```diff
@@ -411,7 +397,18 @@ impl Connection for ConnectHttp {
             return Ok(false);
         }
         let local_channels: HashSet<String> = channels.iter().cloned().collect();
-        let remote_channels: HashSet<String> = HashSet::from_iter(self.channel_list()?);
+        let remote_channels: HashSet<String> = match self.channel_list() {
+            Ok(v) => HashSet::from_iter(v),
+            Err(e) => match e.kind() {
+                UAIDNotRecognizedError(_) => {
+                    // We do not unsubscribe, because the
+                    // server already lost our UAID
+                    self.uaid = None;
```

Comment:
Hmmm that's a good point.. I think you're right. https://autopush.readthedocs.io/en/latest/http.html#registration seems to hint that it'll always return a new UAID **and** `secret`.

Reply:
Okay, added setting `self.auth = None` for now, but this is begging to be refactored. I'll file a follow-up for refactors (including this and the `channel_id` you pointed above!

Reply:
A follow-up for refactors sounds good to me.  I'll just note one issue to tackle: now if `verify_connection` is called twice in a row with a lost UAID, the first one will return `Ok(false)` and the second will be `Err(CommunicationError)`

Reply:
Good point!! Should of caught that - I wonder if we need to fix this right away (happy to do so as a part of this PR!)
a fix would either:
- Keep a new list of (to be re-subscribed) channels in the component, save the state of the component (probably at the `PushManager` level, and return that list directly if any subsequent `verify_connections` get called
- or, keep the persisted state until the consumer calls `subscribe` then wipe the old state. Things can go wrong here if the consumer decides to call something other than `subscribe` so I prefer the first option.

Either way, it depends if we want to support being consistent on two consecutive `verify_connection` @jonalmeida do you happen to know if it's possible for the consumer on android to call a failed `verify_connection` twice before `subscribe`ing? If so I'll add it as a part of this PR, otherwise, I'll do a follow-up PR of that once this merges

Reply:
I'm not sure if this is needed, I don't know enough about how it's used.  But if you want to make that return value consistent, what about reworking how `uaid` and `secret` are stored?  Instead of 2 `Option`s, have a single enum:

```
enum ConnectionState {
  NotConnected,
  NeedsReconnection,
  Connected(uaid: String, auth: String),
}
```

That way `verify_connection` could return the error for `NotConnected` and return `false` for NeedsReconnection.

Reply:
> @jonalmeida do you happen to know if it's possible for the consumer on android to call a failed `verify_connection` twice before `subscribe`ing? If so I'll add it as a part of this PR, otherwise, I'll do a follow-up PR of that once this merges

That shouldn't happen from our current design. Calls to verify_connection, should start the flow that leads to a subscribe event.

Reply:
Thanks Jonathan! @bendk I created https://github.com/mozilla/application-services/issues/4372 to track the reactors you pointed out, since this isn't possible for now we can tackle it there